process-migration-service/src/main/java/org/kie/processmigration/service/impl/MigrationServiceImpl.java

```diff
@@ -0,0 +1,307 @@
+            }
+        } catch (Exception e) {
+            logger.error("Migration [{}] - Callback to {} failed.", migration.getId(), callbackURI, e);
+        }
+    }
+
+    public void validateDefinition(MigrationDefinition definition) throws InvalidMigrationException {
+        if (definition == null) {
+            throw new InvalidMigrationException("The Migration Definition must not be null");
+        }
+        if (definition.getPlanId() == null) {
+            throw new InvalidMigrationException("The Plan ID is mandatory");
+        }
+        if (StringUtils.isBlank(definition.getKieserverId())) {
+            throw new InvalidMigrationException("The KIE Server ID is mandatory");
+        }
+        if (!kieService.getConfigs().containsKey(definition.getKieserverId())) {
+            throw new InvalidKieServerException(definition.getKieserverId());
+        }
+    }
+
+    private void validatePlanExecution(MigrationDefinition definition) throws InvalidMigrationException {
+        try {
+            validatePlanExecution(definition, planService.get(definition.getPlanId()));
+        } catch (PlanNotFoundException e) {
+            throw new InvalidMigrationException("Plan not found with ID: " + definition.getPlanId());
+        }
+    }
+
+    private void validatePlanExecution(MigrationDefinition definition, Plan plan) throws InvalidMigrationException {
+        if (!kieService.existsProcessDefinition(plan.getSourceContainerId(), plan.getTargetProcessId(), definition.getKieserverId())) {
+            throw new ProcessNotFoundException(plan.getSourceContainerId());
+        }
+        if (!kieService.existsProcessDefinition(plan.getTargetContainerId(), plan.getTargetProcessId(), definition.getKieserverId())) {
+            throw new ProcessNotFoundException(plan.getTargetContainerId());
+        }
+    }
+
+    private List<Long> getInstancesToMigrate(Migration migration) throws InvalidKieServerException, PlanNotFoundException {
+        List<Long> instanceIds = migration.getDefinition().getProcessInstanceIds();
+        Plan plan = planService.get(migration.getDefinition().getPlanId());
+        if (instanceIds == null || instanceIds.isEmpty()) {
+            boolean allFetched = false;
+            int page = 0;
+            while (!allFetched) {
+                List<ProcessInstance> instances = kieService.getQueryServicesClient(migration.getDefinition().getKieserverId())
+                                                            .findProcessInstancesByContainerId(plan.getSourceContainerId(), QUERY_PROCESS_INSTANCE_STATUSES, page++, QUERY_PAGE_SIZE);
+
+                instances.stream().forEach(p -> instanceIds.add(p.getId()));
+                if (instances.size() < QUERY_PAGE_SIZE) {
+                    allFetched = true;
+                }
+            }
+        }
+        if (migration.getReports() == null || migration.getReports().isEmpty()) {
```

Comment:
Shouldn't this if be at the beginning of the method? that way we don't have to load instances if report is empty...

Reply:
Good catch!